### PR TITLE
Sync v3

### PR DESCRIPTION
**Pull request by begmaroman**

### CODE REVIEW

File: core/node/rpc/syncv3/service.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	. "github.com/towns-protocol/towns/core/node/protocol"
+	"github.com/towns-protocol/towns/core/node/rpc/syncv3/handler"
+	. "github.com/towns-protocol/towns/core/node/shared"
+)
+
+// Service defines the behavior of the sync V3 service.
+type Service interface {
+	// SyncStreams creates and starts a sync operation. Given streams are immediately going to be added
+	// to the sync operation, and the receiver will receive updates for these streams.
+	SyncStreams(ctx context.Context, id string, streams []*SyncCookie, rec handler.Receiver) error
```
Comment (bas-vk) [inline core/node/rpc/syncv3/service.go:18]: Is it possible to be more precise? In this case it is guaranteed that the caller always receives >= 1 `SyncStreamsResponse` when this function returns nil. It would be nice to desribe these guarantees and expectations concrete.

File: core/node/rpc/syncv3/handler/handler.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+	// DebugDropStream is a debug method to drop a specific stream from the sync operation.
+	// Sends a SyncOp_SYNC_DOWN message to the receiver and unsubscribes from the stream updates.
+	DebugDropStream(ctx context.Context, streamId StreamId) error
+}
+
+// Receiver is a final receiver of the stream update message, i.e. client.
+// It is not thread safe so the race detector will throw an error if multiple goroutines
+// try to call Send at the same time.
+type Receiver interface {
```
Comment (bas-vk) [inline core/node/rpc/syncv3/handler/handler.go:66]: Nitpick, this is an interface and not an implementation. So it's possible that some implementation of this interface are actually thread safe. Better wording is that callers of Send can't make assumptions if Send is thread safe or not.

File: core/node/rpc/service.go [hunk 3]
```diff
@@ -56,8 +58,10 @@ type Service struct {
 	storage         storage.StreamStorage
 
 	// Streams
-	cache       *StreamCache
-	syncHandler river_sync.Handler
+	cache   *StreamCache
+	sync    riversync.Handler
+	syncv3  riversyncv3.Service
+	v3Syncs *xsync.Map[string, struct{}]
```
Comment (bas-vk) [inline core/node/rpc/service.go:64]: Please use better naming than `sync`, `syncv3` and `v3Sync` for fields.

File: core/node/rpc/syncv3/eventbus/eventbus.go [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+	queue    *dynmsgbuf.DynamicBuffer[*eventBusMessage]
+	registry syncer.Registry
+	// subscribers is the list of subscribers grouped by stream ID and syncer version.
+	subscribers map[StreamId]map[int][]StreamSubscriber
+}
+
+// New creates a new instance of the event bus implementation.
+//
+// It creates an internal queue for stream updates and a background goroutine that reads updates from this queue
+// and distributes them to subscribers.
```
Comment (bas-vk) [inline core/node/rpc/syncv3/eventbus/eventbus.go:113]: add `until the given ctx expires.`

File: core/node/rpc/syncv3/eventbus/eventbus.go [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+// It is responsible for handling stream updates from syncer.StreamUpdateEmitter and distributes updates
+// to subscribers. As such it keeps track of which subscribers are subscribed on updates on which streams.
+type eventBusImpl struct {
+	log *logging.Log
+	// TODO: discuss if we want to have an unbounded queue here?
+	// Processing items on the queue should be fast enough to always keep up with the added items.
+	queue    *dynmsgbuf.DynamicBuffer[*eventBusMessage]
+	registry syncer.Registry
+	// subscribers is the list of subscribers grouped by stream ID and syncer version.
+	subscribers map[StreamId]map[int][]StreamSubscriber
```
Comment (bas-vk) [inline core/node/rpc/syncv3/eventbus/eventbus.go:107]: This map must be protected against concurrent access or replaced with a thread-safe alternative. It is modified in `OnStreamEvent` and in a `processXX` methods called in the background task that executes the `run` method.

File: core/node/rpc/syncv3/eventbus/eventbus.go [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+	e.log.Errorw("failed to add stream update message to the queue", "streamID", streamID, "error", err)
+
+	// Unsubscribe event bus from receiving updates of the given stream.
+	e.registry.EnqueueUnsubscribe(streamID)
+
+	// Send sync down message directly to clients to avoid common queue
+	var wg sync.WaitGroup
+	for _, subscribers := range e.subscribers[streamID] {
+		for _, sub := range subscribers {
+			wg.Add(1)
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/node/rpc/syncv3/eventbus/eventbus.go]: Done

File: core/node/rpc/syncv3/eventbus/eventbus.go (hunk 4 above)
Comment (begmaroman) [reply core/node/rpc/syncv3/eventbus/eventbus.go]: Done

File: core/node/rpc/server.go [hunk 7]
```diff
@@ -790,6 +794,19 @@ func (s *Service) initCacheAndSync(opts *ServerStartOpts) error {
...
+		eventBus := eventbus.New(
+			s.serverCtx,
+			s.wallet.Address,
+			s.cache,
+			s.nodeRegistry,
+		)
+		registry := handler.NewRegistry(eventBus)
+		s.syncv3 = syncv3.NewService(registry, s.otelTracer)
+		s.v3Syncs = xsync.NewMap[string, struct{}]()
+	}
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/node/rpc/server.go:805]: Sounds good, done

File: core/node/rpc/syncv3/service.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+	"go.opentelemetry.io/otel/trace"
+
+	. "github.com/towns-protocol/towns/core/node/base"
+	. "github.com/towns-protocol/towns/core/node/protocol"
+	"github.com/towns-protocol/towns/core/node/rpc/syncv3/handler"
+	. "github.com/towns-protocol/towns/core/node/shared"
+)
+
+// Service defines the behavior of the sync V3 service.
+type Service interface {
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/node/rpc/syncv3/service.go:19]: Fixed

File: core/node/rpc/syncv3/handler/handler.go (hunk 2 above)
Comment (begmaroman) [reply core/node/rpc/syncv3/handler/handler.go:64]: Done

File: core/node/rpc/syncv3/eventbus/eventbus.go [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+			go func(sub StreamSubscriber) {
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/node/rpc/syncv3/eventbus/eventbus.go:174]: Worth exploring this option. Added TODO. Let's do it in a future PR as an improvement/optimization.

File: core/node/rpc/service.go (hunk 3 above)
Comment (begmaroman) [reply core/node/rpc/service.go]: Renamed. Not sure if new names are better tho :)

File: core/node/rpc/syncv3/syncer/shared.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+		}
+	}
+}
+
+func (s *sharedStreamUpdateEmitter) StreamID() StreamId {
+	return s.streamID
+}
+
+func (s *sharedStreamUpdateEmitter) Node() common.Address {
+	if s.emitter != nil {
```
Comment (bas-vk) [inline core/node/rpc/syncv3/syncer/shared.go:143]: `s.emitter` is set in the background routine `run` and accessing it here unprotected is a race condition.

File: core/node/rpc/syncv3/handler/handler.go [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+		ctx:           ctx,
+		cancel:        cancel,
+		log:           logging.FromCtx(ctx).Named("syncv3.handler").With("syncID", syncID),
+		syncID:        syncID,
+		receiver:      receiver,
+		eventBus:      s.eventBus,
+		streamUpdates: dynmsgbuf.NewDynamicBuffer[*SyncStreamsResponse](),
+	}
+
+	s.handlers[syncID] = handler
```
Comment (bas-vk) [inline core/node/rpc/syncv3/handler/handler.go:337]: Handlers are never removed causing a memory leak.

File: core/node/rpc/syncv3/syncer/shared.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+				&SyncStreamsResponse{SyncOp: SyncOp_SYNC_DOWN, StreamId: s.streamID[:], TargetSyncIds: br.syncIDs},
+				PendingSubscribersVersion,
+			)
+		}
+		return
+	}
+
+	var emitter StreamUpdateEmitter
+	if stream.IsLocal() {
+		emitter, err = NewLocalStreamUpdateEmitter(
```
Comment (bas-vk) [inline core/node/rpc/syncv3/syncer/shared.go:84]: Because the stream is found you can pass it instead of the stream cache. That saves 1 lookup in the stream cache and gives stricter separations.

File: core/node/rpc/syncv3/syncer/shared.go [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+	s.backfills = append(s.backfills, &backfillRequest{
+		cookie:  cookie,
+		syncIDs: syncIDs,
+	})
+
+	return true
+}
+
+func (s *sharedStreamUpdateEmitter) Close() {
+	if s.emitter != nil {
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/node/rpc/syncv3/syncer/shared.go:177]: Fixed

File: core/node/rpc/syncv3/syncer/shared.go (hunk 12 above)
Comment (begmaroman) [reply core/node/rpc/syncv3/syncer/shared.go:84]: Good point, done.

File: core/node/rpc/syncv3/handler/handler.go (hunk 11 above)
Comment (begmaroman) [reply core/node/rpc/syncv3/handler/handler.go]: Added a logic to remove completed sync operation from both handler registry and event bus.

File: core/node/rpc/syncv3/eventbus/eventbus.go [hunk 14]
```diff
@@ -0,0 +1,561 @@
...
+}
+
+// eventBusImpl is a concrete implementation of the StreamUpdateEmitter and StreamSubscriber interfaces.
+// It is responsible for handling stream updates from syncer.StreamUpdateEmitter and distributes updates
+// to subscribers. As such it keeps track of which subscribers are subscribed on updates on which streams.
+//
+// TODO: Use worker pool to process updates in parallel
+type eventBusImpl struct {
+	log *logging.Log
+	// TODO: discuss if we want to have an unbounded queue here?
```
Comment (sergekh2) [inline core/node/rpc/syncv3/eventbus/eventbus.go:118]: I would say unbounded here is fine. Is there unbounded option on DynamicBuffer

File: core/node/rpc/syncv3/handler/handler.go [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+		return err
+	}
+
+	// Start processing stream updates.
+	var msgs []*SyncStreamsResponse
+	for {
+		select {
+		case <-s.ctx.Done():
+			return s.ctx.Err()
+		case _, open := <-s.streamUpdates.Wait():
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/node/rpc/syncv3/handler/handler.go]: Fixed here and in other places as well.

File: core/node/rpc/syncv3/eventbus/eventbus.go [hunk 16]
```diff
@@ -0,0 +1,561 @@
...
+			delete(e.subscribers, streamID)
+		}
+
+		return
+	}
+
+	// 2. If the version is not PendingSubscribersVersion, it means that the update is sent to subscribers of the stream
+	//    with the given version. We need to find the subscribers for the given version and send the update to them.
+	var wg sync.WaitGroup
+	wg.Add(len(groupedSubscribers[version]))
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/node/rpc/syncv3/eventbus/eventbus.go]: Done

File: core/node/rpc/syncv3/eventbus/eventbus.go [hunk 17]
```diff
@@ -0,0 +1,561 @@
...
+//
+// TODO: Use worker pool to process updates in parallel
+type eventBusImpl struct {
+	log *logging.Log
+	// TODO: discuss if we want to have an unbounded queue here?
+	// Processing items on the queue should be fast enough to always keep up with the added items.
+	queue    *dynmsgbuf.DynamicBuffer[*eventBusMessage]
+	registry syncer.Registry
+	// subscribers is the list of subscribers grouped by stream ID and syncer version.
+	subscribers map[StreamId]map[int][]StreamSubscriber
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/node/rpc/syncv3/eventbus/eventbus.go]: Done

File: core/node/rpc/syncv3/service.go [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+
+	if len(streams) > 0 {
+		res, err := h.Modify(&ModifySyncRequest{SyncId: syncID, AddStreams: streams})
+		if err != nil {
+			_ = h.Cancel(ctx)
+			return err
+		}
+
+		if len(res.GetAdds()) > 0 {
+			return h.Cancel(ctx)
```
Comment (sergekh2) [inline core/node/rpc/syncv3/service.go:83]: unclear why cancel should be called in this case, please add comment.
(and same as above)

File: core/node/rpc/syncv3/service.go [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+	h, err := s.handlerRegistry.New(ctx, syncID, rec)
+	if err != nil {
+		return err
+	}
+	defer s.handlerRegistry.Remove(syncID)
+
+	if len(streams) > 0 {
+		res, err := h.Modify(&ModifySyncRequest{SyncId: syncID, AddStreams: streams})
+		if err != nil {
+			_ = h.Cancel(ctx)
```
Comment (sergekh2) [inline core/node/rpc/syncv3/service.go:78]: Cancel relies on Run loop to dispatch, but Run is not called yet here.

File: core/node/rpc/syncv3/syncer/local.go [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+	l.stream.Unsub(l)
+
+	// Send a stream down message to all active syncs of the current syncer version via event bul.
+	l.subscriber.OnStreamEvent(&SyncStreamsResponse{
+		SyncOp:   SyncOp_SYNC_DOWN,
+		StreamId: l.stream.StreamId().Bytes(),
+	}, l.version)
+
+	// Send a stream down message to all pending syncs, i.e. those that are waiting for backfill.
+	msgs := l.backfillsQueue.GetBatch(nil)
```
Comment (sergekh2) [inline core/node/rpc/syncv3/syncer/local.go:197]: l.backfillsQueue.Close() above drops the buffer, this tries to get the batch, which is alway going to be nil.

If such functionality is desired, CloseAndGetBatch should be added to get the batch atomically.

It seems most of the code would benefit from unit tests. Try using codex with top model to create unit test for corner cases file-by-file. These tests might uncover more edge cases.

File: core/node/rpc/syncv3/syncer/local.go [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+			// nil msgs indicates the buffer is closed.
+			if msgs == nil {
+				l.cancel(nil)
+				return
+			}
+
+			// Messages must be processed in the order they were received.
+			for i, msg := range msgs {
+				// Context could be cancelled while processing messages so adding one more check here.
+				select {
```
Comment (sergekh2) [inline core/node/rpc/syncv3/syncer/local.go:151]: (same as in many other places, lets just replace these with more readable ifs)

File: core/node/rpc/syncv3/syncer/local.go [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+
+	go l.run(ctx)
+
+	return l, nil
+}
+
+// OnUpdate implements events.SyncResultReceiver interface.
+func (l *localStreamUpdateEmitter) OnUpdate(r *StreamAndCookie) {
+	// The first received update should be ignored since all sync ops are starting from their own cookie.
+	if l.firstUpdateReceived.CompareAndSwap(false, true) {
```
Comment (sergekh2) [inline core/node/rpc/syncv3/syncer/local.go:85]: This is fragile, and somewhat unnecessary: refactor stream to add SubNoCookie or such which doesn't send undesired update.

File: core/node/rpc/syncv3/syncer/local.go (hunk 20 above)
Comment (begmaroman) [reply core/node/rpc/syncv3/syncer/local.go]: Added CloseAndGetBatch function.

File: core/node/rpc/syncv3/handler/handler.go [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+// OnUpdate implements the eventbus.StreamSubscriber interface.
+// It is expected to receive SyncOp_SYNC_UPDATE and SyncOp_SYNC_DOWN messages here.
+func (s *syncStreamHandlerImpl) OnUpdate(update *SyncStreamsResponse) {
+	select {
+	case <-s.ctx.Done():
+		return
+	default:
+	}
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/node/rpc/syncv3/handler/handler.go]: Done

File: core/node/utils/dynmsgbuf/dynmsgbuf.go [hunk 24]
```diff
@@ -119,3 +171,29 @@ func (db *DynamicBuffer[T]) Close() {
 	default:
 	}
 }
+
+// CloseAndGetBatch closes the buffer and returns the remaining items.
+func (db *DynamicBuffer[T]) CloseAndGetBatch() []T {
```
Comment (sergekh2) [inline core/node/utils/dynmsgbuf/dynmsgbuf.go:176]: Please modify Close to call CloseAndGetBatch (why have dup code)
